lib/ansible/plugins/action/ce.py

```diff
@@ -0,0 +1,123 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+#
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os
+import sys
+import copy
+
+from ansible.plugins.action.normal import ActionModule as _ActionModule
+from ansible.utils.path import unfrackpath
+from ansible.plugins import connection_loader
+from ansible.module_utils.six import iteritems
+from ansible.module_utils.ce import ce_argument_spec
+from ansible.module_utils.basic import AnsibleFallbackNotFound
+from ansible.module_utils._text import to_bytes
+
+try:
+    from __main__ import display
+except ImportError:
+    from ansible.utils.display import Display
+    display = Display()
+
+
+class ActionModule(_ActionModule):
+
+    def run(self, tmp=None, task_vars=None):
+        if self._play_context.connection != 'local':
+            return dict(
+                failed=True,
+                msg='invalid connection specified, expected connection=local, '
+                    'got %s' % self._play_context.connection
+            )
+
+        provider = self.load_provider()
+        transport = provider['transport'] or 'cli'
+
+        display.vvvv('connection transport is %s' % transport, self._play_context.remote_addr)
+
+        if transport == 'cli':
+            pc = copy.deepcopy(self._play_context)
+            pc.connection = 'network_cli'
+            pc.network_os = 'ce'
+            pc.port = int(provider['port']) or int(self._play_context.port) or 22
+            pc.remote_user = provider['username'] or self._play_context.connection_user
+            pc.password = provider['password'] or self._play_context.password
+            pc.timeout = provider['timeout'] or self._play_context.timeout
+            pc.remote_addr = provider['host'] or self._play_context.remote_addr
+
+            connection = self._shared_loader_obj.connection_loader.get('persistent', pc, sys.stdin)
+            socket_path = self._get_socket_path(pc)
+            display.vvvv('socket_path: %s' % socket_path, pc.remote_addr)
+
+            if not os.path.exists(socket_path):
+                display.vvvv('socket_path not exists: %s' % socket_path, pc.remote_addr)
+                # start the connection if it isn't started
+                rc, out, err = connection.exec_command('open_shell()')
+                display.vvv('open_shell() returned %s %s %s' % (rc, out, err))
+                if rc != 0:
+                    return {'failed': True, 'msg': 'unable to open shell', 'rc': rc}
```

Comment:
We've updated the other action plugins since this PR was raised. If you could diff your code against some of the others and pull in any relevant changes, for example we've improved the error reporting in a few places